velox/functions/prestosql/aggregates/ClassificationAggregation.cpp

```diff
@@ -0,0 +1,720 @@
+          bucketCount_,
+          histogram.bucketCount_);
+    }
+
+    if (min_ != histogram.min_ || max_ != histogram.max_) {
+      VELOX_USER_FAIL(
+          "Cannot merge histograms with different min/max values. "
+          "Left min: {}, left max: {}, right min: {}, right max: {}",
+          min_,
+          max_,
+          histogram.min_,
+          histogram.max_);
+    }
+
+    for (size_t i = 0; i < bucketCount_; ++i) {
+      auto& weights = i >= kMaxBucketCount ? weightsTwo_ : weightsOne_;
+      auto& otherWeights =
+          i >= kMaxBucketCount ? histogram.weightsTwo_ : histogram.weightsOne_;
+      weights[i % kMaxBucketCount] += otherWeights[i % kMaxBucketCount];
+      totalWeights_ += otherWeights[i % kMaxBucketCount];
+    }
+    maxUsedIndex_ = std::max(maxUsedIndex_, histogram.maxUsedIndex_);
+  }
+
+  size_t serializationSize() const {
+    return sizeof(kSerializationVersionHeader) + sizeof(bucketCount_) +
+        sizeof(min_) + sizeof(max_) + (weightsOne_.size() * sizeof(double)) +
+        (weightsTwo_.size() * sizeof(double));
+  }
+
+  /// This represents the total accrued weights in the bucket. The value is
+  /// cached to avoid recomputing it every time it is needed.
+  double totalWeights() const {
+    return totalWeights_;
+  }
+
+ private:
+  /// Returns the index of the bucket in the histogram that contains the
+  /// value. This is done by mapping value to [min, max) and then mapping that
+  /// value to the corresponding bucket.
+  static int64_t
+  getIndexForValue(int64_t bucketCount, double min, double max, double value) {
+    VELOX_CHECK(value >= min && value < max);
+    return std::min(
+        static_cast<int64_t>((bucketCount * (value - min)) / (max - min)),
+        bucketCount - 1);
+  }
+
+  static double getLeftValueForIndex(
+      int64_t bucketCount,
+      double min,
+      double max,
+      int64_t index) {
+    return min + index * (max - min) / bucketCount;
+  }
+
+  static double getRightValueForIndex(
+      int64_t bucketCount,
+      double min,
+      double max,
+      int64_t index) {
+    return std::min(
+        max, getLeftValueForIndex(bucketCount, min, max, index + 1));
+  }
+
+  static void validateParameters(int64_t bucketCount, double min, double max) {
+    if (bucketCount < 2) {
+      VELOX_USER_FAIL("Bucket count must be at least 2.0");
+    }
+
+    if (min >= max) {
+      VELOX_USER_FAIL("Min must be less than max. Min: {}, max: {}", min, max);
+    }
+  }
+
+  static constexpr uint8_t kSerializationVersionHeader = 1;
+  /// In Java, the bucket count is type 'long', which as max possible value of
+  /// 2^63. In C++, a given vector can have max value of upwards of
+  /// std::vector<double>::max_size(), which may be less than 2^63 depending. To
+  /// account for this, we have two buckets which may be used to store the
+  /// weights with each bucket being at most kMaxBucketCount in size.
+  static constexpr int64_t kMaxBucketCount =
```

Comment:
In practice `max_size` is at least 2^60 on 64 bits system, I don't think any system can give that large contiguous memory in one go.  So you don't need to split the weights into 2 arrays for the discrepancy between 2^60 vs 2^64 (you can put a `VELOX_CHECK_LT(bucketCount, weights_.max_size())` in `validateParameters` if you are really concerned about this).

Reply:
Agreed, I added this because it was something the fuzzer caught